bayesian/utility/v2/stack_analyses.py

```diff
@@ -0,0 +1,243 @@
+# Copyright © 2020 Red Hat Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Author: Dharmendra G Patel <dhpatel@redhat.com>
+#
+"""Stack analyses API v2 class, implementing functionality to server POST and GET requests."""
+
+import os
+import datetime
+import uuid
+import json
+
+from requests_futures.sessions import FuturesSession
+from flask import current_app
+
+from sqlalchemy.exc import SQLAlchemyError
+from sqlalchemy.dialects.postgresql import insert
+
+from bayesian import rdb
+from bayesian.exceptions import HTTPError
+from bayesian.utils import (check_for_accepted_ecosystem,
+                            resolved_files_exist,
+                            fetch_sa_request,
+                            retrieve_worker_result)
+from bayesian.dependency_finder import DependencyFinder
+from f8a_worker.models import StackAnalysisRequest
+
+
+worker_count = int(os.getenv('FUTURES_SESSION_WORKER_COUNT', '100'))
+_session = FuturesSession(max_workers=worker_count)
+
+
+class StackAnalyses():
+    """Implements stack analysis API.
+
+    Implements methods to support stack analyses post and get REST APIs calls.
+    """
+
+    def __init__(self, ecosystem=None, manifest=None, file_path=None, show_transitive=None):
+        """Initialize params to be used for ."""
+        self.ecosystem = ecosystem
+        self.manifest = manifest
+        self.file_path = file_path
+        self.show_transitive = show_transitive
+
+        # Private members used during post request.
+        self._new_request_id = None
+        self._request_date_str = None
+        self._manifest_file_info = None
+
+    def post_request(self):
+        """Make stack analyses POST request."""
+        # Make backbone request
+        deps = self._make_backbone_request()
+
+        # Finally save results in RDS and upon success return request id.
+        if self._save_request(deps) == 0:
+            return {
+                "status": "success",
+                "submitted_at": self._date_str,
+                "id": self._new_request_id
+            }
+
+    def validate_params(self):
+        """Validate mandatory post params of the request."""
+        # Manifest is mandatory and must be a string.
+        if self.manifest is None or not resolved_files_exist(self.manifest.filename):
```

Comment:
These can be done before creating an instance of the class. Even if one of them is not present, we do not need to create it.

Reply:
This is not happy case, as most of the API are trigger from plugin, I think saving one object creation is not a big deal. Second, thought of keeping all the spec related stuff inside one class, so that it follows OCP rule.

Reply:
My point is that the validation needs to be done at the point where the variable was initialised. If it is done there then why do you need it here. The other thing is that only the API layer should ideally know about HTTP Errors. By passing it down to the object you are making this unusable from other layers. If you want to do this please raise an exception from here and generate an appropriate HTTP error code on the api layer.

Reply:
Yes, I agree. My new code as moved HTTP error creation in API layer and responder class simply returns code and data.

Reply:
Parameter validation is completely changes now in commit 'Used pydantic to validated post params'. It also takes care of above concern.